source/common/common/perf_annotation.cc

```diff
@@ -32,6 +32,23 @@ void PerfOperation::record(absl::string_view category, absl::string_view descrip
 // because the constructor is non-trivial due to the contained unordered_map.
 PerfAnnotationContext::PerfAnnotationContext() = default;
 
+void PerfAnnotationContext::begin(absl::string_view category, absl::string_view description) {
+  const CategoryDescription key = {std::string(category), std::string(description)};
+  RELEASE_ASSERT(!timestamps_map_.contains(key),
```

Comment:
RELEASE_ASSERT use seems risky.  Are these begin/end annotations meant to be used across event loop events or only within a single call stack?

Reply:
Within a single call stack only unfortunately. Would be nice to have uniquely identifiable events in the event loop though. Is it possible?

Reply:
I think it's OK to use RELEASE_ASSERT as this code will only be activated when the binary is compiled with perf annotation enabled.

And it wouldn't do any good to use ASSERT because you generally would want opt-builds for measuring performance.

Reply:
There's a better way to do all of this: have an RAII object that is kept around for the duration of the perf operation and keeps the necessary state instead of keeping the state on a global as a string in a map.

It nicely solves the issue of concurrent performance measurements and works across call stacks.

Reply:
Yeah for cases where RAII is OK, that works today with `PERF_OPERATION` which is just a wrapper that has all the code compile to nothing when perf annotations are not compiled in. I think that should be preferred when feasible to use.

I assume that PERF_BEGIN and PERF_END are where the plumbing needed to hold an object seems not worth it, and correlating by string seems worth it. It doesn't look to me like this adds any overhead when not used, so that's why I'm OK with this if it makes annotations easier to implement.

Reply:
My bad, I should have given more thought to the question: it may well happen that a "begin" and its corresponding "end" are called for different event loop events e.g. if you put them in ActiveTcpConnection's constructor and destructor respectively.

Reply:
I think that hitting this RELEASE_ASSERT is very likely if the performance measurement ever remains active at the time you unroll the stack to the event loop.  It just takes a second object of that type being benchmarked at the same time.

Reply:
That's true. Applicability of these annotations is quite limited. I mentioned it the doc strings.

Reply:
Sorry I re-read the class and realize that `PerfOperation` does not quite meet the RAII goals @antoniovicente had in mind. That's because in the use-case I built it for, I wanted to start an operation timer but record different categories depending on the outcome of an operation.

A subclass of `PerfOperation` and associated macros could be added where we commit to the category and description at construction time, and the record() happens automatically upon destruction. That would be safer to use than PERF_BEGIN and PERF_END.

@rojkov would that meet your needs? Or do you need the singleton to hold the string-pair map.

Reply:
A RAII based timer is good to measure time spent for a function or a code block. In my case I wanted to measure how exactly wall time is spent during an HTTP request (actually a sequence of many HTTP requests): from establishing a connection to stream creation, from stream creation to headers decoding in a filter, etc... For this I had to maintain a map in the singleton.

Basically I wanted to have a very simple tool for basic [Off-CPU Analysis](http://www.brendangregg.com/offcpuanalysis.html) without resorting to Off-CPU sampling.

Reply:
I haven't fully paged in this conversation, but at minimum doesn't this have to be locked? I'm confused how there wouldn't be race conditions here. Should't we potentially use locking and also allow multiple concurrent measurements that then get added/closed at the end of the sequence? This could be handled by an RAII like handle like @antoniovicente is suggesting?

/wait-any

Reply:
RE locking: yeah that is needed in general; good catch. It might not have been needed by @rojkov 's first usage. Note the ifdef PERF_THREAD_SAFE dictating whether we instantiate and hold the mutexes during map operations. The annotations I had added were (a) on operations that strictly occurred on the main thread and (b) were quick enough that taking the mutex distorted the results a little, hence the ifdef. We should make a PERF_THREAD_SAFE-version of the new map too.

RE RAII, I will let @rojkov describe the use-case in more detail but I thought in some case if you want to start an operation in one scope and end it in another you'd have to manage transfer for the RAII object, and that might be more intrusive to the code you're trying to measure. Using maps allows you to surgically inject some perf visibility without having to have a path to make the object visible in both the place where you are starting and ending the measurement. Having said that, it is not immediately obvious to me why you'd want to measure performance in that scenario.

Reply:
The RAII part doesn't matter, but for this to be useful it seems like we need to support concurrent measurements that might run on multiple workers at the same time?

Reply:
That's a good point; to make this work maybe there could be a third element to the hash-key, which would be a unique ID used for matching in `timestamps_map_`, but which is not used in `duration_stats_map_`, which aggregates across duplicate entries.

Reply:
This PR is a third version of what was done initially. In the first version I mindlessly protected `timestamps_map_` updates with a mutex. [Then](https://github.com/rojkov/envoy/tree/perf-annotations-thread-aware) I realized that the hash-key should better be extended with `std::thread::id` of the current thread.

Finally I stumbled upon what @antoniovicente mentioned earlier: even if there's only one thread it's still possible that a sequence started in e.g. `decodeHeaders()` and stopped in `encodeTrailers()` may stay open waiting for a cluster response when a new connection to the listener triggers an event loop event leading to `PERF_BEGIN()` inside `decodeHeaders()` trying to update the already existing key in the map. That's why I decided to give up thread safety as not really needed for my use case and added some words of warning to the doc strings.

But it would great to extend the hash-key with a third element which could be used universally for matching. I don't know what that could be. Not thread ID.

Reply:
I think the only way it can sensibly work is for the begin API to return a handle (with an internal incrementing ID or whatever) that is used to track the time. In practice, nothing actually has to be added to the perf system until the begin is closed with the end (kind of like the Timespan API we already have). This obviously brings us back to an RAII object that would have to be stored, but IMO this is probably fine as I think with appropriate ifdefs everything can probably be compiled out as a member of an owning object?

Reply:
Yes, the `begin()` could store a `Timespan` object in a global queue then the corresponding `end()` would pop and `complete()` it. Theoretically `end()` might happen to pop a wrong object and that would distort StdDev, but Avg would be correct still.

Reply:
Yeah that imprecision might be OK. But I still think including an extra ID timestamps_map_ may be practical. For example, if you are measuring performance spanning filter callbacks, you could use `this` as your extra ID. The cleanest way though is just for begin() to return an object that you store in your filter undef ifdef; then you don't need a map at all, or to worry about thread safety.

Reply:
> The cleanest way though is just for begin() to return an object that you store in your filter undef ifdef

A-ha, now I see. In this case the owning object (e.g. filter) needs to be annotated too. And it has to be the same for a pair of begin() and end(). Right?

Reply:
Yeah though I think the logical next step of what I suggested is that rather than calling begin() you just instantiate an RAII object yourself, throwing it in an ifdef'd member variable in (say) the filter; maybe via unique_ptr, and destructing it works like end().

Reply:
Understood, thanks!

Reply:
Yes, exactly, ^ is what I'm proposing. I think we can make it so it compiles to nothing if annotations are not enabled.

Reply:
Updated. Turned out it's enough to add a few new macros to achieve the same result.